### ISSUE TITLE
BatchNorm计算结果与torch不同

### ISSUE THREAD

**Issue by SophiaSarah** [PaddlePaddle/Paddle#54080]
### 请提出你的问题 Please ask your question

paddle中bn的计算结果与torch不同，tensor的均值计算、求和、平方结果与torch也不同，float32数据类型，小数点后7位内也有可能不同，如何对齐二者的计算结果？
```
for i in range(5):
    test = np.random.random((batch_size, 80, 52, 52)).astype("float32")
    bn_pd = paddle.nn.BatchNorm2D(80)
    bn_pd.train()
    bn_pd.weight.set_value(np.ones(bn_pd.weight.shape).astype('float32'))
    bn_pd.bias.set_value(np.ones(bn_pd.bias.shape).astype('float32') * 0.1)
    rst_pd = bn_pd((paddle.to_tensor(copy.deepcopy(test))))

    bn_tc = torch.nn.BatchNorm2d(80)
    bn_tc.train()
    torch.nn.init.constant_(bn_tc.weight, 1)
    torch.nn.init.constant_(bn_tc.bias, 0.1)
    rst_tc = bn_tc(torch.from_numpy(copy.deepcopy(test)))
    torch.set_printoptions(16)
    paddle.set_printoptions(16)
    print(rst_pd.shape, rst_tc.shape, rst_pd.min(), rst_tc.min(), rst_pd.max(), rst_tc.max(), rst_pd.mean(), rst_tc.mean(), abs(rst_pd.detach().numpy()-rst_tc.detach().numpy()).max())

for i in range(5):
    test = np.random.random((batch_size, 80, 52, 52)).astype("float32")
    torch.set_printoptions(16)
    paddle.set_printoptions(16)
    rst_pd = (paddle.to_tensor(copy.deepcopy(test))).mean()
    rst_tc = (torch.from_numpy(copy.deepcopy(test))).mean()
    print(rst_pd.shape, rst_tc.shape, rst_pd.dtype, rst_tc.dtype, rst_pd, rst_tc, abs(rst_pd.detach().numpy()-rst_tc.detach().numpy()).max())
    rst_pd = (paddle.to_tensor(copy.deepcopy(test))).sum()
    rst_tc = (torch.from_numpy(copy.deepcopy(test))).sum()
    print(rst_pd.shape, rst_tc.shape, rst_pd.dtype, rst_tc.dtype, rst_pd, rst_tc, abs(rst_pd.detach().numpy() - rst_tc.detach().numpy()).max())
    rst_pd = (paddle.to_tensor(copy.deepcopy(test)))**2
    rst_tc = (torch.from_numpy(copy.deepcopy(test)))**2
    print(rst_pd.shape, rst_tc.shape, rst_pd.dtype, rst_tc.dtype, abs(rst_pd.detach().numpy() - rst_tc.detach().numpy()).max())
```
![image](https://github.com/PaddlePaddle/Paddle/assets/26705571/b8b8b4fe-3330-4533-8697-13330cb6529d)
![image](https://github.com/PaddlePaddle/Paddle/assets/26705571/2bbe69ef-e7b2-4958-a9f4-c0f4117d006d)

**Comment by wangna11BD** [comment]
torch的计算放在cuda上对比一下结果呢

**Comment by SophiaSarah** [comment]
> 类似的情况还是会有，但概率似乎低一些

![image](https://github.com/PaddlePaddle/Paddle/assets/26705571/c078d024-fc4a-4be6-8739-08e46fd5832e)
![image](https://github.com/PaddlePaddle/Paddle/assets/26705571/04ab85c0-cd6e-4ca1-8740-7ee4862e8e13)

**Comment by wangna11BD** [comment]
由于底层选择的算法不同会有一些微小diff，有效数字6位以上相同，我们就认为对齐了的～

**Comment by SophiaSarah** [comment]
> 由于底层选择的算法不同会有一些微小diff，有效数字6位以上相同，我们就认为对齐了的～

了解了，多谢~

**Comment by SophiaSarah** [comment]
> 由于底层选择的算法不同会有一些微小diff，有效数字6位以上相同，我们就认为对齐了的～

想再问下，那这些算法的不同主要涉及哪些算子呢？除BN之外，Conv的计算结果是否可能也不同？

**Comment by SophiaSarah** [comment]
> 由于底层选择的算法不同会有一些微小diff，有效数字6位以上相同，我们就认为对齐了的～
或者说，由于单层计算（比如BN）的结果会与torch存在微小的差异，当这些计算串联，造成的累计误差应该用怎样的标准进行衡量可以认为是对齐的呢？

**Comment by wangna11BD** [comment]
10-6的相对误差累计之后也很小，一般模型最后一层输出相对误差在10-4到10-5就可以认为是对齐的
